src/runtime/test_case_runner.js

```diff
@@ -167,24 +169,38 @@ export default class TestCaseRunner {
   async run() {
     this.emitPrepared()
     this.emit('test-case-started', {})
-    await this.runHooks(
-      this.beforeHookDefinitions,
-      {
-        sourceLocation: this.testCaseSourceLocation,
-        pickle: this.testCase.pickle,
-      },
-      true
-    )
-    await this.runSteps()
-    await this.runHooks(
-      this.afterHookDefinitions,
-      {
-        sourceLocation: this.testCaseSourceLocation,
-        pickle: this.testCase.pickle,
-        result: this.result,
-      },
-      false
-    )
+    let retries = 0
+    for (; retries <= this.retries; retries++) {
+      await this.runHooks(
+        this.beforeHookDefinitions,
+        {
+          sourceLocation: this.testCaseSourceLocation,
+          pickle: this.testCase.pickle,
+        },
+        true
+      )
+      await this.runSteps()
+      await this.runHooks(
+        this.afterHookDefinitions,
+        {
+          sourceLocation: this.testCaseSourceLocation,
+          pickle: this.testCase.pickle,
+          result: this.result,
+        },
+        false
+      )
+      if (this.result.status !== Status.FAILED) {
+        break
+      }
+      if (retries < this.retries) {
+        delete this.result.exception
+        this.result.status = Status.PASSED
+        this.testStepIndex = 0
+      }
+    }
+    if (this.result.status === Status.PASSED && retries > 0) {
+      this.result.status = Status.FLAKY
+    }
```

Comment:
@charlierudolph it seems like my reply was not published, so I'll rewrite it :)

I remember that firing the `test-case-finished` with a `FAILED` status would immediately stop the test, without a chance to re-run it. Having the `FLAKY` status should help now.

From the reporting point of view it still remains the question that formatters (and any report generated out of them) would display multiple runs of the same test case, without the capability of aggregating them.

Reply:
I have moved the two events inside the loop and everything seems fine so far. Just reviewing unit tests for the test case runner.

Reply:
> I remember that firing the test-case-finished with a FAILED status would immediately stop the test, without a chance to re-run it.

That's why I was suggesting a RETRY status (but flakey could work as well). But I want two test cases instead of just one so the formatters can differentiate the runs.

> From the reporting point of view it still remains the question that formatters (and any report generated out of them) would display multiple runs of the same test case, without the capability of aggregating them.

Yes, formatters would need to deal with the multiple runs and aggregate them if they choose or only display the last run of the test case (which I think is what would happen currently)

Reply:
@charlierudolph once I fix the tests I will have a look into the formatters behaviour as well